File: 2017/06/memory-reallocation-2.go
package main

import "fmt"
import "os"
import "io/ioutil"
import "strings"
import "strconv"


func readBanks(data string) (banks []int) {
    fields := strings.Fields(string(data))
    for _, field := range fields {
        bank, err := strconv.Atoi(field)
        if err != nil {
            panic(err)
        }
        banks = append(banks, bank)
    }
    return
}


func writeData(banks []int) string {
    var elements []string
    for _, bank := range banks {
        elements = append(elements, strconv.Itoa(bank))
    }
    return strings.Join(elements, "\t")
}


func redistribute(data string) string {
    banks := readBanks(data)

    // find the largest bank and it's index
    var index, maxBlocks int
    for i, blocks := range banks {
        if blocks > maxBlocks {
            index = i
            maxBlocks = blocks
        }
    }

    // redistrubte the largest bank
    banks[index] = 0
    for maxBlocks > 0 {
        index = (index + 1) % len(banks)
        banks[index] += 1
        maxBlocks -= 1
    }

    return writeData(banks)
}


func countReallocationCycles(banks string) (count int) {
    seen := make(map[string]bool)

    hasKey := false
    for !hasKey {
        seen[banks] = true
        banks = redistribute(banks)
        _, hasKey = seen[banks];
    }

    count = 1
    target := banks
    banks = redistribute(banks)
    for banks != target {
        banks = redistribute(banks)
        count += 1
    }

    return
}


func parseInputFile(filename string) string {
    data, err := ioutil.ReadFile(os.Args[1])
    if err != nil {
        panic(err)
    }
    return writeData(readBanks(string(data)))
}


func main() {
    banks := parseInputFile(os.Args[1])
    count := countReallocationCycles(banks)
    fmt.Println(count)
}

File: 2017/06/memory-reallocation-2.py
# -*- coding: utf-8 -*-
import sys


def redistribute(banks):
    blocks = max(banks)
    index = banks.index(blocks)

    banks = list(banks)
    banks[index] = 0
    for i in range(blocks):
        index = (index + 1) % len(banks)
        banks[index] += 1

    return tuple(banks)


def count_redistribution_cycles(banks):
    seen = set()

    while banks not in seen:
        seen.add(banks)
        banks = redistribute(banks)

    count = 1
    target = banks
    banks = redistribute(banks)
    while banks != target:
        banks = redistribute(banks)
        count += 1

    return count


def main(filename):
    with open(filename) as f:
        data = f.read()

    banks = tuple(int(bank) for bank in data.split())
    return count_redistribution_cycles(banks)


if __name__ == '__main__':
    print(main(sys.argv[1]))
